[HCP Telemetry] Periodic Refresh for Dynamic Telemetry Configuration

Achooo opened the pull request:
### Description

New feature for HCP Telemetry that allows us to dynamically re-configure metrics export to HCP.

**Background** 
Currently, we need to restart the Consul server to bring in any updates to filters, a new endpoint or new labels from the Consul Telemetry Gateway in HCP.

Ideally, we are able to modify these dynamically, which requires logic to continuously fetch new configuration and update it as necessary.

**Implementation** 
This PR enables this functionality, with changes to :
1.  **New object**: `TelemetryConfigProvider`:
   - Periodically fetches telemetry configuration from HCP, using a `time.Ticker`. 
   - Holds configuration and exposes it via methods : `GetEndpoint()`, `GetFilters()` and `GetLabels()`
   - Updates configuration, ONLY if it has changed. Comparisons are done using a hash function on the structs holding the config.
   - Handles concurrency access to read the config and modify the config.
2. **OTELSink**: uses a new `ConfigProvider` interface to access filters and labels.
  - Since labels are now dynamic, we can't add them to the OTEL resource (immutable) so a new function is added to compute OTEL labels dynamically.
  - Handle dynamic filters 
3. **OTELExporter**: uses a new `EndpointProvider` interface to access the endpoint.


** Cleanup / Refactor ***
1.  The `client` package was cleaned to separate out the `TelemetryConfig` object to parse and convert objects. 
2. A logger was added to the client to log bad filter information

This refactor was necessary since now the TelemetryConfigProvider also fetches telemetry config. This avoid duplication of logic for the parsing of the endpoint, labels ,etc.


<!-- Please describe why you're making this change, in plain English. -->

### Testing & Reproduction steps

<!--

* In the case of bugs, describe how to replicate
* If any manual tests were done, document the steps and the conditions to replicate
* Call out any important/ relevant unit tests, e2e tests or integration tests you have added or are adding

-->

- [ ] End to end tests:
- [ ] Unit tests: 

### Links

<!--

Include any links here that might be helpful for people reviewing your PR (Tickets, GH issues, API docs, external benchmarks, tools docs, etc). If there are none, feel free to delete this section.

Please be mindful not to leak any customer or confidential information. HashiCorp employees may want to use our internal URL shortener to obfuscate links.

-->

### PR Checklist

* [x] updated test coverage
* [x] external facing docs updated
* [ ] appropriate backport labels added
* [x] not a security concern

Achooo commented on agent/hcp/client/client.go:
```diff
@@ -329,3 +332,32 @@ func (t *TelemetryConfig) DefaultLabels(cfg config.CloudConfig) map[string]strin
 	return labels
 }
+
+// FilterRegex returns a valid regex used to filter metrics.
```
This was just moved out of the telemetry package from `filter.go`. 

It can be used easily as a method on the `TelemetryConfig`.

For example, in `deps.go`:
https://github.com/hashicorp/consul/pull/18168/files#diff-d652bfc7ebdcf8edacd32ad13fd9059b666981a2a76ef9bc5d985be7edac5601R86-R89


This is specific for our use case so it's a nice cleanup. Someone using filters can provide a regex directly with their own provider.

Achooo commented on agent/hcp/client/client_test.go:
```diff
@@ -199,3 +199,60 @@ func Test_DefaultLabels(t *testing.T) {
 	}
 }
+
+func TestFilterRegex(t *testing.T) {
```
Moved test in here from telemetry package `filter_test.go`

Achooo commented on agent/hcp/deps.go:
The sink and exporter are now initialized with a HCP specific "TelemetryConfigProvider" which continuously fetches configuration from HCP and updates it as needed.

Achooo commented on agent/hcp/telemetry/filter.go:
Moved to client.go

Achooo commented on agent/hcp/telemetry/filter_test.go:
Moved to client_test.go

Achooo commented on agent/hcp/telemetry/otel_exporter.go:
The OTELExporter now holds an `endpointProvider` from which it obtains the endpoint where it needs to send metrics to.

The endpointProvider can be any struct that conforms to the interface. The OTELExporter uses the `GetEndpoint()` of an endpointProvider to obtain the url.

Achooo commented on agent/hcp/telemetry/otel_sink.go:
The OTELSink now holds an `ConfigProvider` from which it obtains the filters and labels to process metrics.

The ConfigProvider can be any struct that conforms to the interface. The OTELSink uses the `GetFilters() and GetLabels()` of an ConfigProvider.

Achooo commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -85,28 +89,19 @@ func NewOTELSink(opts *OTELSinkOpts) (*OTELSink, error) {
+		return nil, fmt.Errorf("ferror: provide valid config provider")
 	}
 
-	attrs := make([]attribute.KeyValue, 0, len(opts.Labels))
```
Don't need labels / filter logic in here since the provider will give this information.

Achooo commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -85,28 +89,19 @@ func NewOTELSink(opts *OTELSinkOpts) (*OTELSink, error) {
+	logger := hclog.FromContext(opts.Ctx).Named("otel_sink")
+
 	// Setup OTEL Metrics SDK to aggregate, convert and export metrics periodically.
-	res := resource.NewWithAttributes("", attrs...)
```
We now use a resource without attributes since labels can change. 

Once created, a resource can't change its attributes. We will now send labels at the metric level instead of OTEL resource level.

jjti replied:
Okay yeah, understood. Payload size will increase, but there shouldn't be any backend change because of this since we wind up adding the attributes to each sample regardless during the OTLP > Prom conversion

Achooo replied:
Sounds good, let me cut a ticket to look into this and gather info. If it does become a problem, I think we should fast track the compression plan that we had previously, which would mean:
1. Consul/collector should compress export payloads
2. TGW should be able to handle compression (think it currently doesn't?)

jjti replied:
> TGW should be able to handle compression (think it currently doesn't?)

Correct that it currently doesn't. There's a trade-off where we'll up our CPU/mem usage in Consul + CTGW but for these repetitive sample labels I imagine it's worth it. When we address this it'd be nice to see how the increased cost of compression helps us w/ payload size.

We use `snappy` encoding already to compress payloads in the gateway, and I see that's a dependency of Consul already, that may be one route (adding `Content-Encoding: snappy`)

chapmanc replied:
Change from http to grpc would result in a better way of handling payload size as well. We should probably backlog this as the HIP client already uses grpc over traefik and we should probably do the same to reduce the egress charges for customers and for our own hcp clusters.

Achooo commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -228,17 +223,30 @@ func (o *OTELSink) flattenKey(parts []string) string {
-	if len(labels) == 0 {
-		return nil
+// labelsToAttributes converts go metrics and provider labels into OTEL format []attributes.KeyValue
+func (o *OTELSink) labelsToAttributes(goMetricsLabels []gometrics.Label) []attribute.KeyValue {
```
We need to handle adding the provider labels at the metric level, in addition to the go metrics label. This function merges both into OTEL attribute key value pairs.

Achooo commented on agent/hcp/telemetry_config_provider.go:
```diff
@@ -0,0 +1,203 @@
+
+// checkUpdate makes a HTTP request to HCP to return a new metrics configuration and true, if config changed.
+// checkUpdate does not update the metricsConfig field to prevent acquiring the write lock unnecessarily.
+func (t *telemetryConfigProvider) checkUpdate(ctx context.Context) (*metricsConfig, bool) {
```
Maybe it would be simpler to do this on the TWG side and return a "hasChanged" field, so that Consul knows there have been changes more easily?

Achooo commented on agent/hcp/telemetry_config_provider_test.go:
I'm still working on these tests:
- need to add a race test
- need to improve the current test. I like the eventually pattern but looking for some insights on this approach

Achooo commented on agent/hcp/client/client.go:
```diff
@@ -87,6 +74,7 @@ func NewClient(cfg config.CloudConfig) (Client, error) {
 
 	client.gnm = hcpgnm.New(client.hc, nil)
 	client.tgw = hcptelemetry.New(client.hc, nil)
+	client.logger = hclog.FromContext(ctx).Named("hcp_client")
```
Small change to add a logger to the client, and obtain it from the context.

Achooo commented on agent/hcp/bootstrap/bootstrap_test.go:
Changes in this file are because I added a logger to the `hcpClient`, which is derived from the context.
A context needs to be passed in when initializing it.
https://github.com/hashicorp/consul/pull/18168/files#r1273942541

Achooo commented on agent/hcp/discover/discover.go:
Changes in this file are because I added a logger to the `hcpClient`, which is derived from the context.
A context needs to be passed in when initializing it.
https://github.com/hashicorp/consul/pull/18168/files#r1273942541

Achooo commented on command/agent/agent.go:
Changes in this file are because I added a logger to the `hcpClient`, which is derived from the context.
A context needs to be passed in when initializing it.
https://github.com/hashicorp/consul/pull/18168/files#r1273942541

Achooo commented on agent/hcp/client/client.go:
Changes:
- Init client with context to add a logger
- Refactored out the `TelemetryConfig`, `MetricsConfig`, new `RefreshConfig`and the conversions into the `telemetry_config.go` file

Why this refactor?
1. This file was growing larger and larger
2. Modified the endpoint, label and filters validation and parsing to be done when converting the response. This reduces the code over in `deps.go` and `telemetry_config_provider.go`. It also avoids code duplication.

Achooo commented on agent/hcp/client/client.go:
```diff
@@ -272,60 +264,3 @@ func (c *hcpClient) DiscoverServers(ctx context.Context) ([]string, error) {
 
 	return servers, nil
 }
-
```
Moved conversions to new file `telemetry_config.go`

Achooo commented on agent/hcp/client/client_test.go:
Removed most tests in this file to put them in `telemetry_config_test.go`
Fixed the actual client test which was not mocking the TGW client properly, and test the validations and conversion.

Achooo commented on agent/hcp/client/telemetry_config.go:
This new file contains all logic pertaining to `TelemetryConfig`, an object that contains telemetry config from HCP.

I settled on the following model for the object: 
```go
type TelemetryConfig struct {
 	MetricsConfig *MetricsConfig
 	RefreshConfig *RefreshConfig
 }
 ```
 In the future, we can add fields here like TraceConfig *TraceConfig or LogsConfig *LogsConfig

I removed the upper level label and endpoint, instead we should perform this conversion and pass around the object **as it should be used**.

This greatly reduces the complexity of code in `deps.go` and `telemetry_config.go`, and removes any duplicated code.

Achooo commented on agent/hcp/deps.go:
Changes:
1. Init the `TelemetryConfigProvider`
2. Remove any parsing that is now unnecessary with the client refactor (url parsing)

Achooo commented on agent/hcp/deps.go:
```diff
@@ -53,50 +56,49 @@ func NewDeps(cfg config.CloudConfig, logger hclog.Logger) (Deps, error) {
 	metricsClient hcpclient.MetricsClient,
-	cfg config.CloudConfig,
-) metrics.MetricSink {
+) (metrics.MetricSink, error) {
```
Now returns an error to test those properly.

Achooo commented on agent/hcp/deps_test.go:
Some of the tests were cleaned up thanks to the refactor of the URL parsing into the client. 
New tests were added for the new TelemetryConfigProvider

Achooo commented on agent/hcp/telemetry_config_provider.go:
New TelemetryConfigProvider which holds dynamic configuration and provides labels, filters and the endpoint to consumers.

Achooo commented on agent/hcp/telemetry_config_provider.go:
```diff
@@ -0,0 +1,182 @@
+
+// run continously checks for updates to the telemetry configuration by making a request to HCP.
+// Modification of config only occurs if changes are detected to decrease write locks that block read locks.
+func (t *telemetryConfigProvider) run(ctx context.Context, refreshInterval time.Duration) {
```
pass in the ctx and refreshInterval to avoid needing a rw lock call here.

Achooo commented on agent/hcp/telemetry_config_provider.go:
```diff
@@ -0,0 +1,182 @@
+	t.cfg = newCfg
+}
+
+// GetEndpoint acquires a read lock to return endpoint configuration for consumers.
```
All the Get* methods acquire a reader lock and defer the unlock before returning the config field

Achooo commented on go.mod:
```diff
@@ -62,7 +62,7 @@ require (
 	github.com/hashicorp/hcl v1.0.0
 	github.com/hashicorp/hcp-scada-provider v0.2.3
-	github.com/hashicorp/hcp-sdk-go v0.48.0
+	github.com/hashicorp/hcp-sdk-go v0.54.0
```
New version to bring in the RefreshConfig changes

Achooo commented on agent/hcp/telemetry_config_provider_test.go:
ASH TODO: Add a race test

Achooo commented on agent/hcp/telemetry/filter.go:
This file was deleted as we parse filters in the client package when creating the `TelemetryConfig` object. The telemetry package shouldn't really care about the filters parsing, especially now as we have a provider

Achooo commented on agent/hcp/telemetry/filter_test.go:
Deleted as tests were moved to `telemetry_config_test.go` in the client package.

Achooo commented on agent/hcp/telemetry_config_provider_test.go:
I'm still working on these tests:

need to add a race test
need to improve the current test
I.e. would it be flaky since we wait to verify the state of the config provider. I can't think of a better way to do this, unless we completely do a unit test on the checkUpdate and reduce the "full test" with time.Ticker. I like the eventually pattern but looking for some insights on this approach

jjti commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -228,17 +222,39 @@ func (o *OTELSink) flattenKey(parts []string) string {
-		return nil
+// filter checks the filter allowlist, if it exists, to verify if this metric should be recorded.
+func (o *OTELSink) allowedMetric(key string) bool {
+	if o.cfgProvider.GetFilters() != nil {
+		return o.cfgProvider.GetFilters().MatchString(key)
```
could `GetFilters()` return a struct on the first call and `nil` on the second? I see it takes out a lock. I wonder if it's config could get updated so it's a nil on second execution. Maybe we want:

```go
if filters := o.cfgProvider.GetFilters(); filters != nil {
...
```

Achooo replied:
oh yep good call thanks!

jjti commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,184 @@
+	"sync"
+	"time"
+
+	goMetrics "github.com/armon/go-metrics"
```
nit: do we need to alias this (and hcpclient, hcpTelemetry below)? I think `metrics` is free (+ looks like the standard in the repo)

double-nit: I like Google's suggestion that aliased imports be named like packages w/o upper-case characters or underscores: https://google.github.io/styleguide/go/decisions#:~:text=including%20the%20prohibition%20on%20the%20use%20of%20underscores%20and%20capital%20letters.

Achooo replied:
fixed!

jjti commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,184 @@
+)
+
+// Ensure hcpProviderImpl implements telemetry provider interfaces.
+var _ hcpTelemetry.ConfigProvider = &hcpProviderImpl{}
+var _ hcpTelemetry.EndpointProvider = &hcpProviderImpl{}
```
this feels a bit coupled to the telemetry configuration in `agent/hcp/client`, do you think this could be moved into there?

Achooo replied:
This was added to force the HCP specific implementation (`hcpProviderImpl`) to conform to those interfaces. Partly documentation and partly ensuring fast at compile time the use case works

Those interfaces will be used by end users to provide their own implementation, so it needs to stay in the telemetry package.

jjti replied:
I understand the confirming to interface check bit, like it, do it a lot

I mean that this implementation is AFAICT the only implementation of those `ConfigProvider` and `EndpointProvider` interfaces. So it seems like this could be in `client` along-side the interfaces?

Achooo replied:
Not sure I fully understand, happy to discuss sync to if that's better! 

Sorry do you mean moving this implementation, the actual implementation `hcpProviderImpl` and files `telemetry_provider*` into `hcp/client` package?

The interfaces, `ConfigProvider` and `EndpointProvider ` are in the `hcp/telemetry` package, and are required for the init of the OTELSink and OTELExporter. The interfaces would have to live there.

jjti replied:
> happy to discuss sync to if that's better!

sounds good to me, would be good to sync tomorrow

> Sorry do you mean moving this implementation, the actual implementation hcpProviderImpl and files telemetry_provider* into hcp/client package?

As you pointed out, I mean `hcp/telemetry`. So yeah, moving this implementation to alongside its interfaces

jjti replied:
We talked and part of the motivation is to keep `hcp/telemetry` free of HCP code so the port to `go-metrics` is easier down the line. If having these implementations outside that package help w/ getting it merged I'm in favor

jjti commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,184 @@
+// NewHCPProviderImpl initializes and starts a HCP Telemetry provider with provided params.
+func NewHCPProviderImpl(ctx context.Context, params *providerParams) (*hcpProviderImpl, error) {
+	if params.hcpClient == nil {
+		return nil, fmt.Errorf("missing HCP client")
```
nit: `errors.New` if not formatting the err

jjti commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,184 @@
+	}
+
+	if params.refreshInterval <= 0 {
+		return nil, fmt.Errorf("invalid refresh interval")
```
```suggestion
		return nil, fmt.Errorf("invalid refresh interval: %d", params.refreshInterval)
```

Achooo replied:
thanks! fixed

jjti commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,184 @@
+	newHash, err := calculateHash(newDynamicConfig)
+	if err != nil {
+		logger.Error("failed to calculate hash for new config", "error", err)
+		goMetrics.IncrCounter(internalMetricRefreshFailure, 1)
```
can we add a success counterpart?

Achooo replied:
done!

jjti commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,184 @@
+// checkUpdate makes a HTTP request to HCP to return a new metrics configuration and true, if config changed.
+// checkUpdate does not update the metricsConfig field to prevent acquiring the write lock unnecessarily.
+func (t *hcpProviderImpl) checkUpdate(ctx context.Context) (*dynamicConfig, bool) {
+	t.rw.RLock()
+	defer t.rw.RUnlock()
```
this lock is taken out to read the existing config? Can we just wrap the grabbing of `t.cfgHash`? This right now wraps the entire network call to HCP which will takes ages in lock terms

Achooo replied:
yep thanks for catching! moved it down

jjti reviewed (commented):
Small comments, suggested changes. Main one being the grabbing of the lock before downloading the Agent config

Achooo commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -23,12 +22,15 @@ import (
 // OTELSinkOpts is used to provide configuration when initializing an OTELSink using NewOTELSink.
 type OTELSinkOpts struct {
-	Reader  otelsdk.Reader
-	Ctx     context.Context
```
After discussion with @chapmanc , preferred to not add ctx in struct so also removed that. It was just used for a logger in the OTELSink

jjti commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,191 @@
+		refreshInterval: telemetryCfg.RefreshConfig.RefreshInterval,
+	}
+
+	newHash, err := calculateHash(newDynamicConfig)
```
we only rarely check the hash (after every network request, which is infrequent). What do you think about, instead of returning `hash` from this func, and storing it on the `hcpProviderImpl`, we instead add a new method `equals` to dynamicConfig like:

```
func (d *dynamicConfig) equals(other *dynamicConfig) bool {
  return hashstructure.Hash(*d, hashstructure.FormatV2, nil) == hashstructure.Hash(*other, hashstructure.FormatV2, nil)
}
```

Small change, but would remove the `hash vars` and the `cfgHash` field

jjti commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,191 @@
+}
+
+// NewHCPProviderImpl initializes and starts a HCP Telemetry provider with provided params.
+func NewHCPProviderImpl(ctx context.Context, params *providerParams) (*hcpProviderImpl, error) {
```
```suggestion
func NewHCPProvider(ctx context.Context, params *providerParams) (*hcpProviderImpl, error) {
```

Personally I wouldn't include `Impl` in the public func. We know it's an implementation of the interface, but it's the only one

jjti commented on agent/hcp/deps.go:
```diff
@@ -53,50 +56,48 @@ func NewDeps(cfg config.CloudConfig, logger hclog.Logger) (Deps, error) {
 
 	telemetryCfg, err := hcpClient.FetchTelemetryConfig(reqCtx)
 	if err != nil {
-		logger.Error("failed to fetch telemetry config", "error", err)
-		return nil
+		return nil, fmt.Errorf("failed to fetch telemetry config: %w", err)
```
I think this creates a situation where if the first call on start up fails we don't start the periodic refresh. The side-effect being that if HCP is down on agent start up, it won't continue querying its config and pick up a valid config when HCP service is restored. My thought is that instead we want a default, fall-back configuration here in case that first fetch fails.

Regarding the default, a couple fields are tricky but doable:
- filters: single wildcard
- labels: need to add cluster, project, org as we do right now (note: this creates a bit of a long-term contract HCP-side)
- endpoint: need to make this based on the environment

jjti replied:
Talked with @Achooo and she proposed a better idea where, rather than guessing defaults, we just return empty values from the config providers. I took a stab at implementing that, and I think it's a pretty small change given existing changes in the PR: https://github.com/hashicorp/consul/pull/18308/files

jjti replied:
If we do this in a follow-up, or decide not to start the sink if the first call fails, I'm fine w/ either. Gonna 🚢 the PR because I'll be OOO tomorrow, don't think this is blocking, and think what's here is a big improvement regardless

jjti reviewed (commented):
Looks good, leaving a few more small suggestions

Achooo commented:
@loshz Working on improving the tests in `telemetry_config_provider.go` to avoid flakiness, but would love some 👀 and feedback earlier so moved to in review.

### Remote Dev Tests

-  Ran Consul docker image on a k8s cluster using consul-k8s
- Used LD flags to update the metrics filters returned by HCP dynamically

We have a a Grafana instance where we can query all metrics exported by a Consul server.
We can see in the Grafana dashboard, new metrics pop in dynamically as I modify filters.

This shows that the `telemetry_config_provider.go` is working as expected.

![Screenshot 2023-07-27 at 12 32 01 PM](https://github.com/hashicorp/consul/assets/23405655/8f054efd-85ce-4e3b-a3a7-abf02060aa42)

jjti commented on agent/hcp/client/telemetry_config.go:
```diff
@@ -0,0 +1,169 @@
+// It fails if there are 0 valid regex filters given.
+func convertMetricFilters(payloadFilters []string) (*regexp.Regexp, error) {
+	var mErr error
+	filters := payloadFilters
```
FWIW this is a 💨 level thought. This matches what we have now, I can see an argument for keeping it as is

Achooo replied:
yep that's actually how the sink was operating (when returning `nil`, allows all metrics), but this is more explicit, I've fixed this to return a default regex  of ".+"

Achooo commented on agent/hcp/manager_test.go:
These are unrelated to my feature work, but the tests kept failing (and sometimes even in my local)

See a build failure: https://github.com/hashicorp/consul/actions/runs/5685610479/job/15411660746

It's because the goroutine never stops when the test exists. 
Once I added the cancel() it seems to have stopped.

chapmanc commented on agent/hcp/deps.go:
```diff
@@ -53,50 +56,44 @@ func NewDeps(cfg config.CloudConfig, logger hclog.Logger) (Deps, error) {
-		Ctx:     ctx,
-		Reader:  telemetry.NewOTELReader(metricsClient, u, telemetry.DefaultExportInterval),
-		Labels:  telemetryCfg.DefaultLabels(cfg),
-		Filters: telemetryCfg.MetricsConfig.Filters,
```
Note: this is now being provided by line 80 the cfgProvider dynamically from the CTGW.

chapmanc commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -38,7 +42,7 @@ type OTELSink struct {
 	// spaceReplacer cleans the flattened key by removing any spaces.
 	spaceReplacer *strings.Replacer
 	logger        hclog.Logger
```
This is fine but nice to not have to have loggers on structs. Unfortunate that go-metrics doesn't pass a context.

Achooo commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,199 @@
+	for {
+		select {
+		case <-tick:
+			if newCfg, hasChanged := t.checkUpdate(ctx); hasChanged {
```
Add a comment here to clarify

Achooo commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,199 @@
+
+// run continously checks for updates to the telemetry configuration by making a request to HCP.
+// Modification of config only occurs if changes are detected to decrease write locks that block read locks.
+func (t *hcpProviderImpl) run(ctx context.Context, tick <-chan time.Time) {
```
remove naming of t

Achooo replied:
Remove the line 96 and put into 102. We don't need to test the ticker functionality as it is a library.

Achooo commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,199 @@
+var _ telemetry.ConfigProvider = &hcpProviderImpl{}
+var _ telemetry.EndpointProvider = &hcpProviderImpl{}
+
+// hasher can be implemented to compute the hash of a dynamicConfig.
```
@loshz : Drop the entire hashing logic and worry about write lock contention.

From a performance perspective, this won't affect much.

Achooo commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,199 @@
+		RefreshInterval: telemetryCfg.RefreshConfig.RefreshInterval,
+	}
+
+	t.rw.RLock()
```
@loshz  : remove hashing , acquire write lock and update cfg

Achooo commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,199 @@
+}
+
+// modifynewTelemetryConfig acquires a write lock to modify it with a given newTelemetryConfig object.
+func (t *hcpProviderImpl) modifyTelemetryConfig(newCfg *dynamicConfig) {
```
This method goes away

Achooo commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,199 @@
+}
+
+// equals returns true if the new dynamicConfig is equal to the current config.
+func (t *hcpProviderImpl) equals(newCfg *dynamicConfig) (bool, error) {
```
This goes away without hashing

Achooo commented on agent/hcp/telemetry/otel_exporter.go:
```diff
@@ -58,7 +69,9 @@ func (e *OTELExporter) Export(ctx context.Context, metrics *metricdata.ResourceM
 	}
-	err := e.client.ExportMetrics(ctx, otlpMetrics, e.endpoint.String())
+
+	endpoint := e.endpointProvider.GetEndpoint()
```
ASH TODO: Validate function

chapmanc reviewed (approved):
LGTM with all the items we have discussed. Thanks for making it iteratively better :)

Achooo commented on agent/hcp/client/telemetry_config_test.go:
```diff
@@ -0,0 +1,383 @@
+			telemetryCfg, err := convertAgentTelemetryResponse(context.Background(), tc.resp, config.CloudConfig{})
+			if tc.wantErr != "" {
+				require.Error(t, err)
+				require.Contains(t, err.Error(), tc.wantErr)
```
Fixed the test to use the error types! they were added in a previous commit. Thanks 🙏🏽

Achooo commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,199 @@
+var _ telemetry.ConfigProvider = &hcpProviderImpl{}
+var _ telemetry.EndpointProvider = &hcpProviderImpl{}
+
+// hasher can be implemented to compute the hash of a dynamicConfig.
```
Resolving as this has been removed

Achooo commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,199 @@
+		select {
+		case <-tick:
+			if newCfg, hasChanged := t.checkUpdate(ctx); hasChanged {
+				t.modifyTelemetryConfig(newCfg)
```
Resolving since we now always update, think that would be too noisy log wise.

Achooo commented on agent/hcp/client/telemetry_config.go:
```diff
@@ -0,0 +1,170 @@
+
+// convertMetricLabels returns a set of <key, value> string pairs that must be added as attributes to all exported telemetry data.
+func convertMetricLabels(payloadLabels map[string]string, cfg config.CloudConfig) map[string]string {
+	labels := make(map[string]string)
```
Yeah I remembered us discussing this, one reason why I think the current solution was chosen is because there is a potential for payloadLabels to be `nil`? i

Achooo commented on agent/hcp/telemetry_provider_test.go:
```diff
@@ -0,0 +1,341 @@
+			wg := &sync.WaitGroup{}
+			// Start goroutines that try to access label configuration.
+			kickOff(wg, testRaceSampleCount, provider, func(provider *hcpProviderImpl) {
+				require.Equal(t, provider.GetLabels(), cfg.MetricsConfig.Labels)
```
Based in our Friday meeting, we decided to let unit tests guarantee behaviour of the state change, and not check the ticker. The `checkUpdate` unit test should guarantee config changes, and this test is just a lock test, that's correct.

The issue is we have then play around with some timing stuff? How do we know the expected updated field? If we initiate an update, we'd have to make sure it's updated once (playing around with timings basically we'd need some sort of mechanism without having direct access to the channels). 

This would have been easy to do with the previous version of the code (with channel access) but I'm not sure if it would be more complex now (manage timings properly etc). If you have any easy ideas, we could pair?


Wonder what

Achooo replied:
I agree it's not the ideal tho, I pivoted mainly based on our last discussion

chapmanc replied:
Im not really concerned with the channel communication but it seems to me like this isn't really testing much. If a single config refresh was successful all of these tests will pass. So if anyone changes your logic (ie changes the refresh logic) how do you guarantee that the readers get the new values. It appears to me at a quick glance that due to the initial load the test would never actually catch if refresh behaviour is changed or running. Do you agree?  This might be okay if you've already tested that in other test scenarios but then I don't really know what the point of this test is? Are you testing a race condition with a race flag?

chapmanc commented on agent/hcp/telemetry_provider.go:
```diff
@@ -0,0 +1,157 @@
+		select {
+		case <-ticker.C:
+			if newCfg := h.getUpdate(ctx); newCfg != nil {
+				ticker.Reset(newCfg.RefreshInterval)
```
might be worth logging the refresh interval but you decide if it's too noisy.

jjti commented:
here or in a follow-up I think we may as well change `sucess` -> `success`. It'd nbd since it's an internal metric, dropping a comment as a bookmark: https://github.com/hashicorp/consul/blob/9ed67e439632f3e29f0b8c1fcf548be7f3588b9d/agent/hcp/telemetry/custom_metrics.go#L9

chapmanc reviewed (approved):
LGTM

loshz reviewed (approved):
LGTM - great work, again!

Achooo commented:
@jjti @loshz @chapmanc 

Added one final tiny commit  : https://github.com/hashicorp/consul/pull/18168/commits/d3522ee86d666157e23a0ab96990e02fcccf4ace

### Fix
I was testing end to end on Consul k8s, and realized:
- `metrics.IncrCounter(internalMetricRefreshSuccess, 1)` custom metric was causing a deadlock. 
- Why? When  `metrics.IncrCounter(internalMetricRefreshSuccess, 1)` is called, under the hood:
  - `FanoutSink -> OTELSink -> TelemetryConfigProvider.GetFilters()` 
  - `GetFilters()` method acquires a read lock  😓 

I've fixed the issue by ensuring we do any custom metric operation, read lock acquisition, BEFORE the write lock is acquired to modify config. I also added a lengthy comment to clarify and ensure anyone is way of this.

### Final E2E tests
I've tested end to end, here is a grafana dashboard showing the change in metrics once I toggle the different metric filters, and how they appear later on.

![Screenshot 2023-08-01 at 3 02 41 PM](https://github.com/hashicorp/consul/assets/23405655/397a0556-8994-4106-a2ae-5539c3de4d32)


I'm now 👍🏽  to merge

Achooo commented on agent/hcp/telemetry_provider.go:
```diff
@@ -120,16 +120,13 @@ func (h *hcpProviderImpl) getUpdate(ctx context.Context) *dynamicConfig {
-	defer h.rw.Unlock()
-
 	h.cfg = newDynamicConfig
+	h.rw.Unlock()
```
@chapmanc recommended to still send the success metric after we update the config in the provider. So moving it below and removing the defer instead.

Achooo commented:
![Screenshot 2023-08-01 at 5 08 43 PM](https://github.com/hashicorp/consul/assets/23405655/eeefcd6c-71ed-4b77-8e38-3e8605c37872)

Final tests conclusive 👍🏽 Also left cluster running for some time and tried modifying the refresh intervals - all looks good. 


Will merge now, thank you all for the great reviews!! 🙏🏽
